bindings/python/mainloop.pyx

```diff
@@ -0,0 +1,17 @@
+
+cdef extern from "sol-mainloop.h":
+	int sol_init()
+	int sol_run()
+	void sol_shutdown()
+
+class Mainloop():
+	def __init__(self):
+		sol_init()
+
+	def __del__(self):
+		sol_shutdown()
```

Comment:
There is `sol_mainloop_set_implementation()`, then when soletta calls `sol_timeout_add()`, you can forward to Python. When Python dispatches, it will call you and you call back C. AFAIU this is similar to what you do in Qt.

The benefit of this approach is that you keep a single entity to manage the process signals, set `sigprocmask()`and spawns threads and forks, each may need some special treatment, as in Soletta we clear mask before proceeding. I recall having issues with that in EFL due the same case of double-management.


that is, in Cython we do something like:

```py

cdef void wrap_sol_init():
    pass

cdef void wrap_sol_shutdown():
   pass

cdef void wrap_sol_quit():
    asyncio.get_event_loop().stop()

cdef void wrap_sol_run():
    asyncio.get_event_loop().run_forever()

cdef wrap_timeout_cb(args):
    cdef object wrap_data = args # may need some cython casts
    handle = wrap_data.pop()
    cdef bool (*cb)(void *data) = wrap_data[0] # may need some cython casts
    cdef void *data = wrap_data[1] # may need some cython casts
    cdef float timeout_s = wrap_data[2]
    if cb(data): # renew
        handle = asyncio.get_event_loop().call_later(timeout_s, wrap_timeout_cb, wrap_data)
        wrap_data.append(handle)

cdef void *wrap_sol_timeout_add(uint32_t timeout_ms, bool (*cb)(void *data), const void *data):
    wrap_data = [<long>cb, <long>data, timeout_ms / 1000.]
    handle = asyncio.get_event_loop().call_later(timeout_ms / 1000., wrap_timeout_cb, wrap_data)
    wrap_data.append(handle)
    return wrap_data # eventually you may need a way to incref it here.

cdef void wrap_sol_timeout_del(void *wrap_handle):
    cdef list wrap_data = wrap_handle # may need some cython casts
    if len(wrap_data) > 3:
        handle = wrap_data.pop()
        handle.cancel() # eventually you may need to decref it here

static const struct sol_mainlooop_implementation _py_asyncio_impl = {
    .init = wrap_sol_init,
    .shutdown = wrap_sol_shutdown,
    .run = wrap_sol_run,
    .quit = wrap_sol_quit,
    .timeout_add = wrap_sol_timeout_add,
    .timeout_del = wrap_sol_timeout_del,
...
}; # I don't recall the syntax!

sol_mainloop_set_implementation(&_py_asyncio_impl) # to be called on module load
```

As said you can use EFL as an example, although it changed a lot since I've stopped contributing, it still looks good to be use as example. They **don't do exactly this mainloop approach!** just use it as example of C<->Cython usage.

A good way to test this is, after the `sol_mainloop_set_implementation()` is called you create a test timer in C/soletta and see if it is fired from asyncio (print or gdb/backtrace):
```py
cdef bool my_test_cb(void *data):
    print("running my test...")
    return 1
sol_timeout_add(1000, my_test_cb, NULL)

```